src/Access/MemoryAccessStorage.cpp

```diff
@@ -201,6 +231,28 @@ bool MemoryAccessStorage::updateNoLock(const UUID & id, const UpdateFunc & updat
 }
 
 
+void MemoryAccessStorage::removeAllExcept(const std::vector<UUID> & ids_to_keep)
+{
+    std::lock_guard lock{mutex};
+    removeAllExceptNoLock(ids_to_keep);
+}
+
+void MemoryAccessStorage::removeAllExceptNoLock(const std::vector<UUID> & ids_to_keep)
+{
+    boost::container::flat_set<UUID> ids_to_keep_set{ids_to_keep.begin(), ids_to_keep.end()};
+    for (auto it = entries_by_id.begin(); it != entries_by_id.end();)
+    {
+        const auto & id = it->first;
+        ++it;
+        if (!ids_to_keep_set.contains(id))
+        {
+            UUID id_to_remove = id;
+            removeNoLock(id_to_remove, /* throw_if_not_exists */ false);
+        }
```

Comment:
This seems a bit dangerous, you delete and iterate at the same time
Why not just collect all ids to delete and call `removeNoLock` on them?

Reply:
`++it` before `removeNoLock()` seems to be enough to stay away of troubles here. Creating a new container of IDs to remove just to pass it better to `removeNoLock()` doesn't seem unnecessary. I added a comment here.